ops/manifests/manifest.py

```diff
@@ -0,0 +1,343 @@
+                if isinstance(manipulate, Patch):
+                    manipulate(rsc)
+
+        return OrderedDict(
+            (HashableResource(obj), None) for obj in all_resources
+        ).keys()
+
+    @lru_cache()
+    def _safe_load(self, filepath: Path) -> List[AnyResource]:
+        """Read manifest file and parse its content into lightkube objects.
+
+        Lightkube can't properly read manifest files which contain List kinds.
+        """
+        content = filepath.read_text()
+        return [
+            codecs.from_dict(item)  # Map to lightkube resources
+            for rsc in yaml.safe_load_all(content)  # load content from file
+            if rsc  # ignore empty objects
+            for item in (rsc["items"] if rsc["kind"] == "List" else [rsc])
+        ]
+
+    def status(self) -> FrozenSet[HashableResource]:
+        """Returns all installed objects which have a `.status.conditions` attribute."""
+        return frozenset(_ for _ in self.installed_resources() if _.status_conditions)
+
+    def installed_resources(self) -> FrozenSet[HashableResource]:
+        """All currently installed resources expected by this manifest."""
+        result: Dict[HashableResource, None] = OrderedDict()
+        for obj in self.resources:
+            try:
+                next_rsc = self.client.get(
+                    type(obj.resource),
+                    obj.name,
+                    namespace=obj.namespace,
+                )
+            except ApiError:
+                continue
+            result[HashableResource(next_rsc)] = None
+        return frozenset(result.keys())
+
+    def labelled_resources(self) -> FrozenSet[HashableResource]:
+        """Any resource ever installed and labeled by this class."""
+        NamespaceKind = namedtuple("NamespaceKind", "namespace, kind")
+        ns_kinds = set(
+            NamespaceKind(obj.namespace, type(obj.resource)) for obj in self.resources
+        )
+
+        return frozenset(
+            HashableResource(rsc)
+            for ns_kind in ns_kinds
+            for rsc in self.client.list(
+                ns_kind.kind,
+                namespace=ns_kind.namespace,
+                labels={
+                    "juju.io/application": self.app_name,
+                    "juju.io/manifest": self.name,
+                },
+            )
+        )
+
+    def apply_manifests(self):
+        """Apply all manifest files from the current release after manipulating."""
+        log.info(f"Applying {self.name} version: {self.current_release}")
+        self.apply_resources(*self.resources)
+
+    def delete_manifests(self, **kwargs):
+        """Delete all manifests associated with the current release."""
+        self.delete_resources(*self.resources, **kwargs)
+
+    def apply_resources(self, *resources: HashableResource):
+        """Apply all manifest files from the current release after manipulating.
+
+        @param *resources: set of resourecs to apply, if empty apply all
```

Comment:
where is the logic for doing the apply all case when its empty?

Reply:
ooh... comment mistake :)